packages/block-library/src/query/index.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+/**
+ * Server-side rendering of the `core/query` block.
+ *
+ * @package WordPress
+ */
+class Blocks_Query {
+	/**
+	 * @var array
+	 */
+	public static $displayedPostIds = array();
+}
+
+/**
+ * Registers the `core/query` block on server.
+*/
+function register_block_core_query() {
+	$path = __DIR__ . '/query/block.json';
+	$metadata = json_decode( file_get_contents( $path), true);
+	register_block_type(
+		$metadata['name'],
+		array_merge(
+			$metadata,
+			array(
+				'render_callback' => 'render_block_core_query',
+			)
+		)
+	);
+}
+
+/**
+ * Convert criteria object into args ready for use in WP_Query
+ *
+ * @param array $criteria A criteria object.
+ *
+ * @return array Return an array of args.
+ */
+function core_query_attributes_to_critera( $criteria ) {
+	if ( $criteria[ 'specificMode' ] == 1 ) {
+		$args = array(
+			'post_status' => 'publish',
+			'p' => $criteria[ 'singleId' ],
+		);
+		return $args;
+	}
+
+	$args = array(
+		'posts_per_page'      => ! empty( $criteria['per_page'] ) ? intval( $criteria['per_page'] ) : 3,
+		'post_status'         => 'publish',
+		'suppress_filters'    => false,
+		'ignore_sticky_posts' => true,
+	);
+
+	if ( ! empty( $criteria['author'] ) ) {
+		$args['author'] = implode( ",", $criteria['author'] );
+	}
+	if ( ! empty( $criteria['categories'] ) ) {
+		$args['cat'] = implode( ",", $criteria['categories'] );
+	}
+	if ( ! empty( $criteria['tags'] ) ) {
+		$args['tag_in'] = intval( $criteria['tags'] );
+	}
+	if ( ! empty( $criteria['search'] ) ) {
+		$args['s'] = sanitize_text_field( $criteria['search'] );
+	}
+	return $args;
+}
+
+function render_block_core_query( $attributes ) {
+	$blocks = ! empty( $attributes['blocks'] ) ? $attributes['blocks'] : array();
+	$args = core_query_attributes_to_critera( $attributes['criteria'] );
+	$args['posts_per_page'] = $args['posts_per_page'] + count( Blocks_Query::$displayedPostIds );
+
+	$query = new WP_Query( $args );
```

Comment:
What happens if we just want to use the default query object? We should probably use the default query if no arguments are present?

Reply:
Are you changing the default WP_Query or just looking to make sure that this does what you'd expect?

I think that getting this to align with WP_Query is a good goal, but for now there are a lot of options that don't have a UI yet. This is something I think we'll have to iterate on.

Reply:
I'm just envisioning using this in a block-based theme. I'd add the block in my theme but not add any custom query args so that it will use the global WP_Query object. I don't want to trigger the server to do a second query if it isn't needed. 

I could be misunderstanding, but this seems like the foundational use case for a block based theme?

Reply:
> I could be misunderstanding, but this seems like the foundational use case for a block based theme?

If it isn't, I don't know what is. I just tried out the new block-based theming system and the loop appears to be the only thing missing. The template tag blocks need to be placed within some kind of loop container, they can't just be directly placed in the outer part of the template. Especially for posts page, archives and search results templates where the default query returns multiple posts, the template needs to have some kind of loop container and this appears to be it.